File: server/replay_server_test.py
import base64
import json
import tempfile
import threading
import unittest
import urllib.error
import urllib.request
from http.server import ThreadingHTTPServer
from pathlib import Path

from replay_server import initialize_database, make_handler


class ReplayServerTest(unittest.TestCase):
    def setUp(self):
        self.temporary_directory = tempfile.TemporaryDirectory()
        database_path = str(Path(self.temporary_directory.name) / "scores.sqlite3")
        initialize_database(database_path)
        self.server = ThreadingHTTPServer(("127.0.0.1", 0), make_handler(database_path))
        self.thread = threading.Thread(target=self.server.serve_forever, daemon=True)
        self.thread.start()
        self.url = f"http://127.0.0.1:{self.server.server_port}"

    def tearDown(self):
        self.server.shutdown()
        self.server.server_close()
        self.thread.join()
        self.temporary_directory.cleanup()

    def request(self, path, data=None):
        body = None if data is None else json.dumps(data).encode()
        request = urllib.request.Request(
            self.url + path,
            data=body,
            headers={"Content-Type": "application/json"},
        )
        return urllib.request.urlopen(request)

    def test_submit_list_and_download_replay(self):
        replay = b"compressed replay bytes"
        submission = {
            "chart_id": "chart-1",
            "nickname": "Nimue",
            "mode": "mania",
            "score": 123456,
            "accuracy": 0.987,
            "replay": base64.b64encode(replay).decode(),
        }
        with self.request("/scores", submission) as response:
            self.assertEqual(response.status, 201)
            score_id = json.load(response)["id"]

        with self.request("/leaderboard?chart_id=chart-1") as response:
            scores = json.load(response)["scores"]
            self.assertEqual(scores[0]["nickname"], "Nimue")
            self.assertEqual(scores[0]["accuracy"], 0.987)

        with self.request(f"/scores/{score_id}/replay") as response:
            self.assertEqual(response.read(), replay)

    def test_defaults_to_anonymous_and_orders_by_score(self):
        for score in (10, 20):
            self.request(
                "/scores",
                {
                    "chart_id": "chart-1",
                    "score": score,
                    "replay": base64.b64encode(b"replay").decode(),
                },
            ).close()

        with self.request("/leaderboard?chart_id=chart-1") as response:
            scores = json.load(response)["scores"]
            self.assertEqual([score["score"] for score in scores], [20, 10])
            self.assertEqual(scores[0]["nickname"], "Anonymous")


if __name__ == "__main__":
    unittest.main()

File: server/replay-server.test.mjs
import assert from "node:assert/strict";
import { afterEach, beforeEach, test } from "node:test";
import { openReplayDatabase, createReplayServer } from "./replay-server.mjs";

let database;
let server;
let base_url;

beforeEach(async () => {
  database = openReplayDatabase(":memory:");
  server = createReplayServer({ database });
  await new Promise((resolve) => server.listen(0, "127.0.0.1", resolve));
  base_url = `http://127.0.0.1:${server.address().port}/api`;
});

afterEach(async () => {
  await new Promise((resolve) => server.close(resolve));
  database.close();
});

async function request(path, options = {}) {
  const response = await fetch(`${base_url}${path}`, options);
  const result = await response.json();
  return { response, result };
}

async function auth(path, name, password = "secret1") {
  return request(path, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ name, password }),
  });
}

async function submit({ token, accuracy, score = 0,
  chart_md5 = "11111111111111111111111111111111", chart_index = 1, mode = "mania" }) {
  const result = await request("/scores", {
    method: "POST",
    headers: { "Content-Type": "application/json", ...(token ? { Authorization: `Bearer ${token}` } : {}) },
    body: JSON.stringify({ chart_md5, chart_index, mode, accuracy, score,
      played_at: new Date().toISOString(), replay: "AAEC" }),
  });
  assert.equal(result.response.status, 201, JSON.stringify(result.result));
  return result;
}

test("registers and logs in a case-insensitive user", async () => {
  const registration = await auth("/register", "Nimue");
  assert.equal(registration.response.status, 201);
  assert.equal(registration.result.user.name, "Nimue");
  assert.ok(registration.result.token);

  const duplicate = await auth("/register", "nimue");
  assert.equal(duplicate.response.status, 409);
  const login = await auth("/login", "NIMUE");
  assert.equal(login.response.status, 200);

  const me = await request("/me", { headers: { Authorization: `Bearer ${login.result.token}` } });
  assert.equal(me.result.user.name, "Nimue");
});

test("shows only a registered user's best chart score and every anonymous score", async () => {
  const { result: registration } = await auth("/register", "Player");
  await submit({ token: registration.token, accuracy: 0.8, score: 800 });
  await submit({ token: registration.token, accuracy: 0.95, score: 950, mode: "osu" });
  await submit({ accuracy: 0.9, score: 900 });
  await submit({ accuracy: 0.7, score: 700 });

  const { result } = await request("/leaderboard?chart_md5=11111111111111111111111111111111&chart_index=1");
  assert.deepEqual(result.scores.map((play) => [play.nickname, play.accuracy, play.registered]), [
    ["Player", 0.95, true],
    ["Anonymous", 0.9, false],
    ["Anonymous", 0.7, false],
  ]);
});

test("lists recent plays newest first without deduplicating users", async () => {
  const { result: registration } = await auth("/register", "Player");
  await submit({ token: registration.token, accuracy: 0.8 });
  await submit({ accuracy: 0.9, chart_md5: "22222222222222222222222222222222" });

  const { result } = await request("/scores/recent?limit=2");
  assert.deepEqual(result.scores.map((play) => [play.chart_md5, play.chart_index, play.nickname]), [
    ["22222222222222222222222222222222", 1, "Anonymous"],
    ["11111111111111111111111111111111", 1, "Player"],
  ]);
});
